### DISCUSSION TITLE
Cannot compile iOS Xcode Project

### DISCUSSION THREAD

**Discussion by Wattnco** [flyinghead/flycast#712] [Q&A]
Hi there,

Problem 1:
I am trying to compile the Xcode project using /GitHub/flycast/shell/apple/generate_xcode_project.command and it seems to keep complaining about missing stuff which I cannot seem to correct. It does make the Xcode project but I am worried it wont compile properly?

-- Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES) 
-- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES) 
-- Could NOT find OpenMP (missing: OpenMP_C_FOUND OpenMP_CXX_FOUND)

I have installed libomp through both brew and macports and still get this error.

-- Could NOT find ALSA (missing: ALSA_LIBRARY ALSA_INCLUDE_DIR) 

Alsa is not available via macports (so I installed portaudio) and brew says "alsa-lib: Linux is required for this software."

`/Users/nic/Documents/GitHub/flycast/shell/apple/generate_xcode_project.command ; exit;
1) macOS (arm64)
2) iOS
Choose your target platform: 2
Building iOS xcodeproj for debugging
Remove CODE_SIGNING_ALLOWED=NO in Build Settings if you are using your Apple Developer Certificate for signing
-- The C compiler identification is AppleClang 14.0.0.14000029
-- The CXX compiler identification is AppleClang 14.0.0.14000029
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The Swift compiler identification is Apple 5.7
-- Check for working Swift compiler: /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swift-frontend
-- Check for working Swift compiler: /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swift-frontend - works
-- Found Git: /usr/bin/git (found version "2.37.0 (Apple Git-136)") 
-- Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES) 
-- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES) 
-- Could NOT find OpenMP (missing: OpenMP_C_FOUND OpenMP_CXX_FOUND) 
-- Performing Test CFLAG_Wall
-- Performing Test CFLAG_Wall - Success
-- Performing Test CFLAG_Wextra
-- Performing Test CFLAG_Wextra - Success
-- Performing Test CFLAG_Wcast_qual
-- Performing Test CFLAG_Wcast_qual - Success
-- Performing Test CFLAG_Wcast_align
-- Performing Test CFLAG_Wcast_align - Success
-- Performing Test CFLAG_Wshadow
-- Performing Test CFLAG_Wshadow - Success
-- Performing Test CFLAG_Wstrict_aliasing_1
-- Performing Test CFLAG_Wstrict_aliasing_1 - Success
-- Performing Test CFLAG_Wswitch_enum
-- Performing Test CFLAG_Wswitch_enum - Success
-- Performing Test CFLAG_Wdeclaration_after_statement
-- Performing Test CFLAG_Wdeclaration_after_statement - Success
-- Performing Test CFLAG_Wstrict_prototypes
-- Performing Test CFLAG_Wstrict_prototypes - Success
-- Performing Test CFLAG_Wundef
-- Performing Test CFLAG_Wundef - Success
-- Could NOT find ALSA (missing: ALSA_LIBRARY ALSA_INCLUDE_DIR) 
-- Found ZLIB: /Applications/Xcode-beta.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.0.sdk/usr/lib/libz.tbd (found version "1.2.11") 
-- Could NOT find Lua (missing: LUA_LIBRARIES LUA_INCLUDE_DIR) 
-- Found PkgConfig: /opt/local/bin/pkg-config (found version "0.29.2") 
-- Looking for _close
-- Looking for _close - not found
-- Looking for _dup
-- Looking for _dup - not found
-- Looking for _fdopen
-- Looking for _fdopen - not found
-- Looking for _fileno
-- Looking for _fileno - not found
-- Looking for _setmode
-- Looking for _setmode - not found
-- Looking for _snprintf
-- Looking for _snprintf - not found
-- Looking for _strdup
-- Looking for _strdup - not found
-- Looking for _stricmp
-- Looking for _stricmp - not found
-- Looking for _strtoi64
-- Looking for _strtoi64 - not found
-- Looking for _strtoui64
-- Looking for _strtoui64 - not found
-- Looking for _unlink
-- Looking for _unlink - not found
-- Looking for arc4random
-- Looking for arc4random - found
-- Looking for clonefile
-- Looking for clonefile - found
-- Looking for explicit_bzero
-- Looking for explicit_bzero - not found
-- Looking for explicit_memset
-- Looking for explicit_memset - not found
-- Looking for fileno
-- Looking for fileno - found
-- Looking for fseeko
-- Looking for fseeko - found
-- Looking for ftello
-- Looking for ftello - found
-- Looking for getprogname
-- Looking for getprogname - found
-- Looking for localtime_r
-- Looking for localtime_r - found
-- Looking for setmode
-- Looking for setmode - found
-- Looking for strcasecmp
-- Looking for strcasecmp - found
-- Looking for strdup
-- Looking for strdup - found
-- Looking for stricmp
-- Looking for stricmp - not found
-- Looking for strtoll
-- Looking for strtoll - found
-- Looking for strtoull
-- Looking for strtoull - found
-- Looking for 3 include files sys/types.h, ..., fts.h
-- Looking for 3 include files sys/types.h, ..., fts.h - found
-- Looking for fts_open
-- Looking for fts_open - found
-- Looking for include file stdbool.h
-- Looking for include file stdbool.h - found
-- Looking for include file strings.h
-- Looking for include file strings.h - found
-- Looking for include file unistd.h
-- Looking for include file unistd.h - found
-- Looking for include file inttypes.h
-- Looking for include file inttypes.h - found
-- Looking for include file stdint.h
-- Looking for include file stdint.h - found
-- Looking for include file sys/types.h
-- Looking for include file sys/types.h - found
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of __int8
-- Check size of __int8 - failed
-- Check size of int8_t
-- Check size of int8_t - done
-- Check size of uint8_t
-- Check size of uint8_t - done
-- Check size of __int16
-- Check size of __int16 - failed
-- Check size of int16_t
-- Check size of int16_t - done
-- Check size of uint16_t
-- Check size of uint16_t - done
-- Check size of __int32
-- Check size of __int32 - failed
-- Check size of int32_t
-- Check size of int32_t - done
-- Check size of uint32_t
-- Check size of uint32_t - done
-- Check size of __int64
-- Check size of __int64 - failed
-- Check size of int64_t
-- Check size of int64_t - done
-- Check size of uint64_t
-- Check size of uint64_t - done
-- Check size of short
-- Check size of short - done
-- Check size of int
-- Check size of int - done
-- Check size of long
-- Check size of long - done
-- Check size of long long
-- Check size of long long - done
-- Check size of off_t
-- Check size of off_t - done
-- Check size of size_t
-- Check size of size_t - done
-- Performing Test HAVE_FICLONERANGE
-- Performing Test HAVE_FICLONERANGE - Failed
-- Performing Test HAVE_NULLABLE
-- Performing Test HAVE_NULLABLE - Success
-- Found ZLIB: /Applications/Xcode-beta.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.0.sdk/usr/lib/libz.tbd (found suitable version "1.2.11", minimum required is "1.1.2") 
CMake Warning at core/deps/libzip/CMakeLists.txt:224 (message):
  -- neither Common Crypto, GnuTLS, mbed TLS, OpenSSL, nor Windows
  Cryptography found; AES support disabled

-- Found Perl: /opt/local/bin/perl (found version "5.34.1") 
-- Configuring done
-- Generating done
CMake Warning (dev) at core/deps/libzip/lib/CMakeLists.txt:5 (add_library):
  Policy CMP0063 is not set: Honor visibility properties for all target
  types.  Run "cmake --help-policy CMP0063" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  Target "zip" of type "STATIC_LIBRARY" has the following visibility
  properties set for C:

    C_VISIBILITY_PRESET

  For compatibility CMake is not honoring them for this target.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Build files have been written to: /Users/nic/Documents/GitHub/flycast/build

Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.

[Process completed]
`

Problem 2
Once I open the Xcode project I get the following errors:

- Flycast 1 issue " error build: Command Ld failed with a nonzero exit code"

The project builds fine if I choose build for running. But when I try to archive it I get the same error:

- Flycast 1 issue " error build: Command Ld failed with a nonzero exit code"

How can I fix this?

Thanks!

**Comment by vkedwardli** [comment]
- Problem 1 `Could NOT find OpenMP `
OpenMP is not supported in iOS (yet, I haven't spent time to investigate this)
The libomp you installed from brew is compiled for macOS but not iOS
OpenMP is used for Texture Upscaling

- Problem 2 `Could NOT find ALSA`
you can ignore this also

- Problem 3 `Command Ld failed with a nonzero exit code`
Please expand the error and see what is it trying to link and failed?
I just tried the `generate_xcode_project.command` on an Intel Mac, and can build an ipa successfully
I guess I need to use an Apple Silicon Mac to test then

**Comment by vkedwardli** [comment]
Just test on Apple Silicon, using Xcode 13.2.1, I can build the ipa successfully also

**Reply by Wattnco** [reply]
Hmm I wonder what I am doing wrong then?

When I build and try to run on my iPhone I get a successful build but it gives an install error:

`Details

Unable to install "Flycast"
Domain: com.apple.dt.MobileDeviceErrorDomain
Code: -402620388
User Info: {
    DVTErrorCreationDateKey = "2022-08-09 22:59:37 +0000";
    IDERunOperationFailingWorker = IDEInstalliPhoneLauncher;
}
--
No code signature found.
Domain: com.apple.dt.MobileDeviceErrorDomain
Code: -402620388
User Info: {
    DVTRadarComponentKey = 364477;
    MobileDeviceErrorCode = "(0xE800801C)";
    "com.apple.dtdevicekit.stacktrace" = (
	0   DTDeviceKitBase                     0x000000010ff9c2cc DTDKCreateNSErrorFromAMDErrorCode + 300
	1   DTDeviceKitBase                     0x000000010ffd03d0 __90-[DTDKMobileDeviceToken installApplicationBundleAtPath:withOptions:andError:withCallback:]_block_invoke + 136
	2   DVTFoundation                       0x0000000105b3a8ec DVTInvokeWithStrongOwnership + 76
	3   DTDeviceKitBase                     0x000000010ffd0154 -[DTDKMobileDeviceToken installApplicationBundleAtPath:withOptions:andError:withCallback:] + 1196
	4   IDEiOSSupportCore                   0x000000010f54fdbc __118-[DVTiOSDevice(DVTiPhoneApplicationInstallation) processAppInstallSet:appUninstallSet:installOptions:completionBlock:]_block_invoke.147 + 2328
	5   DVTFoundation                       0x0000000105c44b44 __DVT_CALLING_CLIENT_BLOCK__ + 16
	6   DVTFoundation                       0x0000000105c455b0 __DVTDispatchAsync_block_invoke + 152
	7   libdispatch.dylib                   0x00000001b15c4a48 _dispatch_call_block_and_release + 32
	8   libdispatch.dylib                   0x00000001b15c6570 _dispatch_client_callout + 20
	9   libdispatch.dylib                   0x00000001b15cdc28 _dispatch_lane_serial_drain + 672
	10  libdispatch.dylib                   0x00000001b15ce774 _dispatch_lane_invoke + 384
	11  libdispatch.dylib                   0x00000001b15d934c _dispatch_workloop_worker_thread + 652
	12  libsystem_pthread.dylib             0x00000001b1749100 _pthread_wqthread + 288
	13  libsystem_pthread.dylib             0x00000001b1747e20 start_wqthread + 8
);
}
--

Analytics Event: com.apple.dt.IDERunOperationWorkerFinished : {
    "device_model" = "iPhone13,4";
    "device_osBuild" = "15.6 (19G71)";
    "device_platform" = "com.apple.platform.iphoneos";
    "launchSession_schemeCommand" = Run;
    "launchSession_state" = 1;
    "launchSession_targetArch" = arm64;
    "operation_duration_ms" = 2426;
    "operation_errorCode" = "-402620388";
    "operation_errorDomain" = "com.apple.dt.MobileDeviceErrorDomain";
    "operation_errorWorker" = IDEInstalliPhoneLauncher;
    "operation_name" = IDEiPhoneRunOperationWorkerGroup;
    "param_consoleMode" = 0;
    "param_debugger_attachToExtensions" = 0;
    "param_debugger_attachToXPC" = 1;
    "param_debugger_type" = 5;
    "param_destination_isProxy" = 0;
    "param_destination_platform" = "com.apple.platform.iphoneos";
    "param_diag_MainThreadChecker_stopOnIssue" = 0;
    "param_diag_MallocStackLogging_enableDuringAttach" = 0;
    "param_diag_MallocStackLogging_enableForXPC" = 1;
    "param_diag_allowLocationSimulation" = 1;
    "param_diag_gpu_frameCapture_enable" = 0;
    "param_diag_gpu_shaderValidation_enable" = 0;
    "param_diag_gpu_validation_enable" = 0;
    "param_diag_memoryGraphOnResourceException" = 0;
    "param_diag_queueDebugging_enable" = 1;
    "param_diag_runtimeProfile_generate" = 0;
    "param_diag_sanitizer_asan_enable" = 0;
    "param_diag_sanitizer_tsan_enable" = 0;
    "param_diag_sanitizer_tsan_stopOnIssue" = 0;
    "param_diag_sanitizer_ubsan_stopOnIssue" = 0;
    "param_diag_showNonLocalizedStrings" = 0;
    "param_diag_viewDebugging_enabled" = 1;
    "param_diag_viewDebugging_insertDylibOnLaunch" = 1;
    "param_install_style" = 0;
    "param_launcher_UID" = 2;
    "param_launcher_allowDeviceSensorReplayData" = 0;
    "param_launcher_kind" = 0;
    "param_launcher_style" = 0;
    "param_launcher_substyle" = 0;
    "param_runnable_appExtensionHostRunMode" = 0;
    "param_runnable_productType" = "com.apple.product-type.application";
    "param_runnable_type" = 2;
    "param_testing_launchedForTesting" = 0;
    "param_testing_suppressSimulatorApp" = 0;
    "param_testing_usingCLI" = 0;
    "sdk_canonicalName" = "iphoneos16.0";
    "sdk_osVersion" = "16.0";
    "sdk_variant" = iphoneos;
}
--


System Information

macOS Version 13.0 (Build 22A5321d)
Xcode 14.0 (21322) (Build 14A5284g)
Timestamp: 2022-08-10T08:59:37+10:00`


I have changed the signing to be automatic and selected my dev account and changed the bundle identifier to com.MYBUSINESSNAME.flycast.

Is there anything else that I am missing that would cause the "no code signature found" error?

Thanks for taking time to help me with this!

**Reply by Wattnco** [reply]
PS no idea why that is all formatted in bold...

**Reply by Wattnco** [reply]
Ok I got this to build via Xcode by removing the "CODE_SIGNING=NO" line from the Cmake file before making the the project via the .command file. 

So it now installs on my iPhone yay!

But...

Now I get an error when trying to open the bios or a game:

00:11:051 hw/mem/_vmem.cpp:494 N[VMEM]: Info: nvmem is enabled, with addr space of size 4GB
00:11:072 hw/mem/_vmem.cpp:593 N[VMEM]: BASE 0x2a8010000 RAM(16 MB) 0x334010000 VRAM64(8 MB) 0x32c010000 ARAM(2 MB) 0x328810000
00:11:082 hw/mem/_vmem.cpp:494 N[VMEM]: Info: nvmem is enabled, with addr space of size 4GB
00:11:086 hw/mem/_vmem.cpp:593 N[VMEM]: BASE 0x2a8010000 RAM(16 MB) 0x334010000 VRAM64(8 MB) 0x32c010000 ARAM(2 MB) 0x328810000
00:11:197 rend/gui.cpp:2740 E[COMMON]: Verify Failed  : found
 in ngen_Rewrite -> /Users/nic/Github_Local/flycast/core/rec-ARM64/rec_arm64.cpp : 2280

Not sure how to fix that?
